src/AutoSelect.ml

```diff
@@ -0,0 +1,209 @@
+  inherit nopCilVisitor
+
+  method! vinst = function
+    | Call (_,Lval ((Var info), NoOffset),args,_,_) as call->
+      (*ignore @@ Pretty.fprint stdout 50 (printInstr defaultCilPrinter () call) ;*)
+      callSet := FunctionSet.add info !callSet;
+      calledBy := FunctionCallMap.update info (addOrCreateMap fd) !calledBy;
+      (*We collect the argument list so we can use LibraryFunctions.find to classify functions*)
+      argLists := FunctionCallMap.add info args !argLists;
+      (*print_endline @@ fd.vname ^ " -> " ^ info.vname;
+      Pretty.fprint stdout (Pretty.d_list "\n" (fun () e -> printExp defaultCilPrinter () e) () args) ~width:50;
+      print_endline "\n";*)
+      DoChildren
+    | _ -> DoChildren
+end;;
+
+class functionVisitor(calling, calledBy, argLists) = object
+  inherit nopCilVisitor
+  
+    method! vfunc fd = 
+      let callSet = ref FunctionSet.empty in
+      let callVisitor = new collectFunctionCallsVisitor (callSet, calledBy, argLists, fd.svar) in
+      ignore @@ Cil.visitCilFunction callVisitor fd;
+      calling := FunctionCallMap.add fd.svar !callSet !calling;
+      (*print_endline fd.svar.vname;
+      ignore (dumpGlobal plainCilPrinter stdout (GFun (fd, !currentLoc)));*)
+      SkipChildren
+  end;;
+
+let functionCallMaps = ResettableLazy.from_fun (fun () ->
+  let calling = ref FunctionCallMap.empty in
+  let calledBy = ref FunctionCallMap.empty in
+  let argLists = ref FunctionCallMap.empty in
+  let thisVisitor = new functionVisitor(calling,calledBy, argLists) in
+  visitCilFileSameGlobals thisVisitor (!Cilfacade.current_file);
+  !calling, !calledBy, !argLists);;
+
+(*TODO Extend to dynamic calls?*)
+let calledFunctions fd = ResettableLazy.force functionCallMaps |> fun (x,_,_) -> x |> FunctionCallMap.find_opt fd |> Option.value ~default:FunctionSet.empty;;
+let callingFunctions fd = ResettableLazy.force functionCallMaps |> fun (_,x,_) -> x |> FunctionCallMap.find_opt fd |> Option.value ~default:FunctionSet.empty;;
+let functionArgs fd = ResettableLazy.force functionCallMaps |> fun (_,_,x) -> x |> FunctionCallMap.find_opt fd;;
+
+(*Functions for determining if the Congruence analysis should be enabled *)
+let isNotExtern = function
+  | Extern -> false
+  | _ -> true;;
+
+let rec setCongruenceRecursive fd depth neigbourFunction =
+  if depth >= 0 then (
+    fd.svar.vattr <- addAttributes (fd.svar.vattr) [Attr ("goblint_precision",[AStr "congruence"])];
+    FunctionSet.iter 
+      (fun vinfo -> 
+        print_endline ("    " ^ vinfo.vname);
+        setCongruenceRecursive (Cilfacade.find_varinfo_fundec vinfo) (depth -1) neigbourFunction
+      ) 
+      (FunctionSet.filter 
+        (fun x -> isNotExtern x.vstorage) 
+        (neigbourFunction fd.svar) 
+      )
+    ;
+  );;
+
+(*Because we now ignore the total count and only care if it exists, this should be rewritten to stop after the first mod*)
+(* -> No longer a visitor, but a recursive search?*)  
```

Comment:
A visitor would still be much shorter since you don't have to rewrite all the structural recursion over the AST. You can use a custom exception to jump out of the visitor and catch the exception outside, which is a fairly common OCaml programming technique for this sort of thing.